Mitigate Jacoco report generation issues

snazy opened the pull request:
Externalizes the logic to wait for the Jacoco data file into `DataFileWatch`, which should handle the cases when the data file does not already exist (aka `JacococConfig.reuseDataFile==false`) and if it already exists (aka `JacocoConfig.reuseDataFile==true`).

`DataFileWatch` remembers the "initial file size", which is the file size _before_ Jacoco runs. If the file does not already exists, the file size is `-1L`. The implementation waits basically "inifinitely" as long as the file size of the data file changes, which is interpreted as "data file is being updated by Jacoco". If the data file size stays at its "initial file size" for 10 seconds and the Jacoco shutdown hook's not running, the wait's aborted. If the "wait for Jacoco data file" has timed out, an additional message is printed to `stderr`.

Also changes the "jacoco running" detection from inspecting the running threads to checking whether the Jacoco MBean is (still) registered.

If the "wait for Jacoco data file" has timed out, an additional message is printed to `stderr`.

The previous code also registered one shutdown-hook to create the report from the Jacoco data file for every _test case_ being executed. This lead to `number-of-test-cases` shutdown hooks to generate the same Jacoco report from the same Jacoco data file. The correct behavior is to generate one report when the JVM terminates. This is implemented by setting a system property, because every _test class_ results in a new Quarkus class loader running a different class instance of `JacocoProcessor`, using a static field in `JacocoProcessor` would still cause `number-of-test-classes` shutdown hooks.

- Fixes: #30264

snazy committed:
Fix Jacoco report generation

Externalizes the logic to wait for the Jacoco data file into `DataFileWatch`, which should handle the cases when the data file does not already exist (aka `JacococConfig.reuseDataFile==false`) and if it already exists (aka `JacocoConfig.reuseDataFile==true`).

`DataFileWatch` remembers the "initial file size", which is the file size _before_ Jacoco runs. If the file does not already exists, the file size is `-1L`. The implementation waits basically "inifinitely" as long as the file size of the data file changes, which is interpreted as "data file is being updated by Jacoco". If the data file size stays at its "initial file size" for 10 seconds and the Jacoco shutdown hook's not running, the wait's aborted. If the "wait for Jacoco data file" has timed out, an additional message is printed to `stderr`.

Also changes the "jacoco running" detection from inspecting the running threads to checking whether the Jacoco MBean is (still) registered.

If the "wait for Jacoco data file" has timed out, an additional message is printed to `stderr`.

The previous code also registered one shutdown-hook to create the report from the Jacoco data file for every _test case_ being executed. This lead to `number-of-test-cases` shutdown hooks to generate the same Jacoco report from the same Jacoco data file. The correct behavior is to generate one report when the JVM terminates. This is implemented by setting a system property, because every _test class_ results in a new Quarkus class loader running a different class instance of `JacocoProcessor`, using a static field in `JacocoProcessor` would still cause `number-of-test-classes` shutdown hooks.

Fixes: #30264

famod commented:
> and the Jacoco shutdown hook's not running

Just a quick heads up (I haven't checked your code in detail yet), but [the code we already have to detect whether the hook is still running](https://github.com/quarkusio/quarkus/blob/2.15.3.Final/test-framework/jacoco/runtime/src/main/java/io/quarkus/jacoco/runtime/ReportCreator.java#L57) doesn't seem to work reliably: https://github.com/quarkusio/quarkus/issues/18571#issuecomment-905032697

snazy commented:
> the code we already have to detect whether the hook is still running doesn't seem to work reliably

I suspect that's the loop that checks the running thread stack traces? I left that piece, not sure whether it's actually useful. I'm still running tests with this change - luckily no failure so far.

dimas-b commented on test-framework/jacoco/runtime/src/main/java/io/quarkus/jacoco/runtime/DataFileWatch.java:
```diff
@@ -0,0 +1,81 @@
+        // either the "abort timeout" is hit or the jacoco thread is no longer running.
+        long abortTime = nextAbortTime();
+        long fileSize;
+        for (long previousFileSize = initialDataFileSize;; previousFileSize = fileSize) {
```
nit: add a space between `;;` for readability?

dimas-b replied:
... or perhaps convert to `while(true)`?

snazy replied:
Nope, code's automagically formatted

famod commented:
> I suspect that's the loop that checks the running thread stack traces? I left that piece, not sure whether it's actually useful.

Yes, that bit. I fear we might always run into a ~10s delay on systems/setups where that check doesn't work reliably.

dimas-b commented on test-framework/jacoco/runtime/src/test/java/io/quarkus/jacoco/runtime/DataFileWatchTest.java:
```diff
@@ -0,0 +1,229 @@
+        DataFileWatch dataFileWatch = new DataFileWatch(clock::get, datafile, jacocoFinished::get) {
+            @Override
+            void waitSleep() {
+                switch (waitIterations.decrementAndGet()) {
```
Just a suggestion: Use a `CyclicBarrier` or a `Semaphore` to wait/flip between the watch thread and the test thread and control side effects from the test thread in a simple top-down sequential code block.

dimas-b reviewed (commented):
The overall approach looks solid to me. I hope it actually fixes coverage data collection :) ... a couple of minor comments below.

snazy commented:
> Yes, that bit. I fear we might always run into a ~10s delay on systems/setups where that check doesn't work reliably.

Good point!
So the jacoco shutdown hook's thread class is an anonymous inner class extending `j.l.Thread` (in `o.j.agent.rt.internal.Agent#getInstance()`) - i.e. it _should_ work, but as you said, there's no guarantee.
However, if we enable JMX we could check the existence of the `org.jacoco:type=Runtime` mbean, which feels safer to check ; it's deregistered after the data file's been written.

dimas-b commented on test-framework/jacoco/runtime/src/main/java/io/quarkus/jacoco/runtime/DataFileWatch.java:
```diff
@@ -0,0 +1,81 @@
+                    // Give the jacoco thread writing the data file some time to write remaining data.
+                    abortTime = nextAbortTime();
+                }
+                if (fileSize > 0L && jacocoFinished.getAsBoolean()) {
```
The `jacocoFinished` check can give a false negative, but not a false positive (as long as jacoco's `Thread` sub-class remains the same). 

What do you think about finding the JaCoCo thread up front and using `Thread.join(timeout)` and/or `getState()` to get notifications of JaCoCo finishing instead?

dimas-b replied:
If we monitor the JaCoCo MBean, should the `fileSize > 0L` check be `fileSize > initialDataFileSize` to cover the case of reusing (appending) the data file?

snazy replied:
It's implicitly checked via the outer `if (fileSize != initialDataFileSize)` . I'm not sure whether the data file will be always growing (not that familiar with the data file format at all).

snazy commented:
Updated the jacoco-running check to use JMX and also added some (rather very ugly) code to only add the report-creation shutdown-hook once per test JVM. I'm not familiar how quarkus extensions really work, so I'm not sure whether there's a better way than the added `JacocoProcessor.transformedClasses` field - or whether that's necessary at all.

dimas-b commented on test-framework/jacoco/runtime/src/main/java/io/quarkus/jacoco/runtime/ReportCreator.java:
```diff
@@ -111,12 +97,23 @@ public void run() {
 
+    private static boolean jacocoRunning() {
+        try {
+            ManagementFactory.getPlatformMBeanServer().getObjectInstance(new ObjectName("org.jacoco:type=Runtime"));
```
Do we really need the file size check now?

snazy replied:
Yes, the chance that we miss the MBean being there is pretty high (shutdown-hooks run concurrently).

geoand commented:
@snazy what is the status of this?

snazy commented:
@geoand I think, this one's ready for review. My local tests all passed w/o hitting the `EOFException`. The only thing that I'm unsure about is whether the "cached `BytecodeTransformerBuildItem`s" that I added help or whether it can be removed.

geoand commented:
@famod any chance you can give this a try with your setup as well?

famod commented:
I try and have a look this evening.

geoand commented:
Thanks!

famod commented:
Got tangled up in [#30400](https://github.com/quarkusio/quarkus/issues/30400), still need to test this.

famod commented:
So I tested this a bit:
- Normally, I don't actually use the report feature but rely on EclEmma (Eclipse) and a Jenkins plugin, but I enabled it for my tests.
- I haven't seen any EOFExceptions, neither in IDE nor on the command line (I have to say though that back then I did _not always_ get the EOFException in Eclipse).
- When running a test in Eclipse I do get matching coverage data in the report.
- When running a test in Eclipse _with coverage_ (EclEmma) I do _not_ get matching coverage data in the report.
  Instead I get that new delay and warning about the missing exec file (which is not there indeed).
- When running a test (or all in a module) via `mvn` (jacoco-agent is always active), then I can see a very noticeable delay after the last test, but I don't see the message (this could be a known bug in surefire-plugin; output from shutdown hooks is lost). 
  Unsurprisingly, the report doesn't show any coverage.
  What's strange here: The `.exec` file _does exist_! Might be worth mentioning that I do _not_ use `append`.

So it's a mixed bag. But here is a twist: In 2.15.3, in the same cases as above, I don't get coverage data in report either!
The way I see it with the new delay, it's now even more relevant to disable the report feature when it isn't working for you anyway (and/or you get your reports differently, like me).
We could advice users to disable the report (by extending the message), but 
[SUREFIRE-1879](https://issues.apache.org/jira/browse/SUREFIRE-1879) might swallow it anyway.

geoand commented:
Thanks for the detailed report!

Given your experience, I am not really sure how to proceed

famod commented:
> I am not really sure how to proceed

Same here.

Some late in the evening brainstorming:

Option 1: We merge it (after a more detailed code review), with an extended message (regardless of SUREFIRE-1879) and add something to the migration guide and/or docs.

Option 2: We keep the old impl but add this new approach as an alternative and let users decide (via config). In this case there should also be some kind of documentation/communication so that we get sufficient feedback on what to keep in the long run.

Option 3: We try to explore other ways.

snazy commented:
I've updated the PR a little bit:
* error messages, also those from the shutdown hook, are written to `error.txt`, so there's at least some chance that users can notice those
* updated the file-resolve mechanism a bit - maybe the `dataFile` field does in some situations point to a different file (wild guess, out of thin air)

I'm not sure whether the report generation from `JacocoProcessor`, having two "racing" shutdown hooks, is great. We need ways to know whether Jacoco ran / is still active, and whether the data file was changed.

The change to use JMX is probably fine, and maybe it's possible to update Jacoco a little to help fixing this one:
* Add a new method `boolean isShutdown()` to `org.jacoco.agent.rt.IAgent` - or maybe even some `waitForShutdown()`
* Add a third value to the `agent.jmx` option: `keep-alive` (in addition to `false` + `true`) to not unregister the mbean on shutdown

Naively speaking, it could be an option to remove the report generation from `JacocoProcessor` and let the Maven/Gradle plugins deal with the report generation, but that would add Jacoco stuff to the build tool plugins, which is probably not such a great option.

snazy commented:
(Jacoco change in mind: https://github.com/jacoco/jacoco/compare/master...snazy:jacoco:extend-agent-quarkus?expand=1)

geoand commented:
@famod would you be able to retest?

famod commented:
Sorry for the delay. I'll try and have a look this week.

geoand commented:
Thanks 🙏

gsmet reviewed (dismissed):
Just a quick note that this will need a squash at the end before merging.
Please dismiss once done.

cescoffier commented:
@snazy Sorry... needs another rebase :-(

cescoffier commented:
@snazy can you rebase?
